From fc295ad46ad1a4c6a2f3d40ba42689f5c951978c Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Tue, 5 Mar 2019 18:42:32 +0530
Subject: [PATCH 197/255] octeontx2-pf: Fix memory leak due to error packets

commit b40e0c50692fcc2df90be25cfb2f90b7fd8946a5 from
git@git.assembla.com:cavium/WindRiver.linux.git

When HW adds a CQE_RX for a received error packet it also
consumes a buffer, ignoring that will lead to memory leak
and also stalls packet reception as number of buffers in
Aura/Pool will come down with each error pkt.

This patch fixes this by freeing the IOVA back to Aura.

Change-Id: I2b065d51aaf059aac4123e7b4105fde685d22b9a
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../ethernet/marvell/octeontx2/nic/otx2_txrx.c    | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_txrx.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_txrx.c
index 0cfe43059a10..fcb0c7e3a28a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_txrx.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_txrx.c
@@ -194,13 +194,6 @@ static void otx2_rcv_pkt_handler(struct otx2_nic *pfvf,
 
 	/* CQE_HDR_S for a Rx pkt is always followed by RX_PARSE_S */
 	parse = (struct nix_rx_parse_s *)(cqe + sizeof(*cqe_hdr));
-	/* Check for errors */
-	if (parse->errlev || parse->errcode) {
-		dev_info(pfvf->dev,
-			 "RQ%d: Error pkt received errlev %x errcode %x\n",
-			 cq->cint_idx, parse->errlev, parse->errcode);
-		return;
-	}
 
 	start = cqe + sizeof(*cqe_hdr) + sizeof(*parse);
 	end = start + ((parse->desc_sizem1 + 1) * 16);
@@ -225,6 +218,14 @@ static void otx2_rcv_pkt_handler(struct otx2_nic *pfvf,
 		iova = (void *)sg + sizeof(*sg);
 
 		for (seg = 0; seg < sg->segs; seg++) {
+			/* Check for errors */
+			if (parse->errlev || parse->errcode) {
+				otx2_aura_freeptr(pfvf, cq->cq_idx,
+						  *iova & ~0x07ULL);
+				iova++;
+				continue;
+			}
+
 			len = sg_lens[frag_num(seg)];
 			/* Starting IOVA's 2:0 bits give alignment
 			 * bytes after which packet data starts.
-- 
2.17.1

